### ISSUE TITLE
Weird roadblocking loop-closure bug

### ISSUE THREAD

**Issue by phaikawl** [gopherjs/gopherjs#163]
http://pastebin.com/vLsY8CQA
Result should be `1 2 3` (with native Go)
gopherjs returns `3 3 3` instead.

**Comment by shurcooL** [comment]
http://play.golang.org/p/BEcxCKsc4S outputs `1 2 3`.
http://www.gopherjs.org/play/#/BEcxCKsc4S outputs `3 3 3`.

**Comment by eanderton** [comment]
That playground link isn't showing all of the generated code for some reason.  By moving the definition f() into an anonymous func() within main(), I got this:

http://www.gopherjs.org/play/#/1S_DsH6Lta

Looks like the anonymous functions generated in the loop are not closing over 'z' correctly.  This smells like how JavaScript handles these closures, where the idiom is to pass in a loop variable in order to ensure that a copy is closed over.

**Comment by eanderton** [comment]
I simplified the playground code to cut closer to the problem at hand:

http://www.gopherjs.org/play/#/W8aWXHhqdT

Sure enough, the JS code closes over 'v', but not 'the value of v', which is the expected behavior.  GopherJS probably needs to either pass in all closed-over vars into a shim that generates the anonymous function:

```
fn = function(v) {
return function() {
  do_something_with_v(v);
};
}(v);
```

**Comment by neelance** [comment]
@eanderton Thanks for your effort. However it seems like you simplified it a bit too much, because for your test case it has the correct behaviour, see http://play.golang.org/p/w3nk9d8c_c . The range variable in Go is not in a new scope for each iteration. That's one of the not so nice behaviours of Go and I hope that it will be fixed with Go 2.0.

**Comment by eanderton** [comment]
@neelance: Whoops. Sorry about that.
